domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,470 @@
+      = runBlockingTest(coroutineContext) {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    shadowMediaPlayer.invokeCompletionListener()
+    advanceTimeBy(2000)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.COMPLETED)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_observeData_removeObserver_checkJobIsNull()
+      = runBlockingTest(coroutineContext) {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observeForever(mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    playProgress.removeObserver(mockAudioPlayerObserver)
+    advanceTimeBy(2000)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_addAndRemoveObservers_checkJobStates() = runBlockingTest(coroutineContext) {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+
+    playProgress.observeForever(mockAudioPlayerObserver)
+    audioPlayerController.pause()
+
+    playProgress.removeObserver(mockAudioPlayerObserver)
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+  }
+
+  @Test
+  fun testController_alreadyInitialized_initializePlayer_fails() {
+    arrangeMediaPlayer()
+
+    assertThrows(IllegalStateException::class) {
+      audioPlayerController.initializeMediaPlayer(TEST_URL2)
+    }
+  }
+
+  @Test
+  fun testController_notInitialized_releasePlayer_fails() {
+    assertThrows(IllegalStateException::class) {
+      audioPlayerController.releaseMediaPlayer()
+    }
+  }
+
+  @Test
+  fun testError_notPrepared_invokePlay_fails() {
+    assertThrows(IllegalStateException::class) {
+      audioPlayerController.play()
+    }
+  }
+
+  @Test
+  fun testError_notPrepared_invokePause_fails() {
+    assertThrows(IllegalStateException::class) {
+      audioPlayerController.pause()
+    }
+  }
+
+  @Test
+  fun testError_notPrepared_invokeSeekTo_fails() {
+    assertThrows(IllegalStateException::class) {
+      audioPlayerController.seekTo(500)
+    }
+  }
+
+  private fun arrangeMediaPlayer() {
+    audioPlayerController.initializeMediaPlayer(TEST_URL).observeForever(mockAudioPlayerObserver)
+    shadowMediaPlayer.invokePreparedListener()
+  }
+
+  private fun addMediaInfo() {
+    val dataSource = DataSource.toDataSource(context , Uri.parse(TEST_URL))
+    val mediaInfo = ShadowMediaPlayer.MediaInfo(/* duration= */ 1000,/* preparationDelay= */ 0)
+    ShadowMediaPlayer.addMediaInfo(dataSource, mediaInfo)
+  }
+
+  private fun <T: Throwable> assertThrows(type: KClass<T>, operation: () -> Unit): T {
```

Comment:
Please copy over the TODO from the other test with assertThrowsSync to move this to a common test infrastructure.